coreax/least_squares.py

```diff
@@ -0,0 +1,342 @@
+    """
+    Given an array's eigendecomposition, return the inverse of the array.
+
+    .. warning::
+        We assume the order of the ``eigenvalues`` and ``eigenvectors`` correspond, i.e.
+        the first element of ``eigenvalues`` is paired with the first column of
+        ``eigenvectors``.
+
+    :param eigenvalues: Vector of eigenvalues
+    :param eigenvectors: Array of eigenvectors as columns
+    :param rcond: Cut-off ratio for small eigenvalues
+    :return: Approximate inverse of array using its eigendecomposition
+    """
+    # Mask the eigenvalues that are zero or almost zero according to value of rcond
+    # for safe inversion.
+    mask = eigenvalues >= jnp.array(rcond) * jnp.max(eigenvalues)
+    safe_eigenvalues = jnp.where(mask, eigenvalues, 1)
+
+    # Invert the eigenvalues safely and extend array for broadcasting
+    inverse_eigenvalues = jnp.where(mask, 1 / safe_eigenvalues, 0)[:, jnp.newaxis]
+
+    # Solve Ax = I, x = A^-1 = UL^-1U^T
+    return eigenvectors.dot(inverse_eigenvalues * eigenvectors.T)
+
+
+class RandomisedEigendecompositionSolver(RegularisedLeastSquaresSolver):
+    """
+    Approximate solution to regularised linear equations via random eigendecomposition.
+
+    Solving regularised linear equations involving large arrays can become prohibitively
+    expensive as size increases. To reduce this computational cost, the solution can be
+    approximated by various methods. :class:`RandomisedEigendecompositionSolver`
+    is a class that does such an approximation using the randomised eigendecomposition
+    of the input array.
+
+    .. warning::
+        Input arrays must be Hermitian for this method to have predictable
+        behaviour. We do not check this.
+
+    Using Algorithm 4.4. and 5.3 from :cite:`halko2009randomness` we approximate the
+    eigendecomposition of a Hermitian matrix. The parameters `oversampling_parameter`
+    and `power_iterations` present a trade-off between speed and approximation quality.
+    See :cite:`halko2009randomness` for discussion on choosing sensible parameters; the
+    defaults chosen here are cautious.
+
+    :param random_key: Key for random number generation
+    :param oversampling_parameter: Number of random columns to sample; the larger the
+        oversampling_parameter, the more accurate, but slower the method will be
+    :param power_iterations: Number of power iterations to do; the larger the
+        power_iterations, the more accurate, but slower the method will be
+    :param rcond: Cut-off ratio for small singular values of the ``array``. For the
+        purposes of rank determination, singular values are treated as zero if they are
+        smaller than rcond times the largest singular value of a. The default value of
+        :data:`None` will use the machine precision multiplied by the largest dimension
+        of the ``array``. An alternate value of -1 will use machine precision
```

Comment:
Reinstate full stop.

For multi-sentence descriptions, we keep the final full stop, otherwise ditch it. This is why I prefer single sentence, resorting to semicolons if necessary. This description is so large that it ought to be multi-sentence.